core/src/main/java/overflowdb/BatchedUpdate.java

```diff
@@ -0,0 +1,330 @@
+        public Object src;
+        public Object dst;
+        public Object[] propertiesAndKeys;
+
+        public CreateEdge(String label, Object src, Object dst, Object[] propertiesAndKeys) {
+            this.label = label;
+            this.src = src;
+            this.dst = dst;
+            this.propertiesAndKeys = propertiesAndKeys;
+        }
+    }
+
+    public static class SetNodeProperty implements Change{
+        public String label;
+        public Node node;
+        public Object value;
+
+        public SetNodeProperty(String label, Node node, Object value) {
+            this.label = label;
+            this.node = node;
+            this.value = value;
+        }
+    }
+
+    public static AppliedDiff applyDiff(Graph graph, DiffOrBuilder diff){
+        return new DiffGraphApplier(graph, diff, null, null).run();
+    }
+
+    public static AppliedDiff applyDiff(Graph graph, DiffOrBuilder diff, KeyPool keyPool, ModificationListener listener){
+        return new DiffGraphApplier(graph, diff, keyPool, listener).run();
+    }
+
+
+
+    private static class DiffGraphApplier{
+        private final DiffOrBuilder diff;
+        private final KeyPool keyPool;
+        private final ModificationListener listener;
+        private final ArrayDeque<DetachedNodeData> deferredInitializers = new ArrayDeque<>();
+        private final Graph graph;
+        private int nChanges = 0;
+
+        DiffGraphApplier(Graph graph, DiffOrBuilder diff, KeyPool keyPool, ModificationListener listener) {
+            this.diff = diff;
+            this.keyPool = keyPool;
+            this.listener = listener;
+            this.graph = graph;
+        }
+
+        AppliedDiff run(){
+            try {
+                for (Iterator<Change> it = diff.iterator(); it.hasNext(); ) {
+                    Change change = it.next();
+                    applyChange(change);
+                }
+            } finally {
+                if(listener != null)
+                    listener.finish();
+            }
+            return new AppliedDiff(graph, diff, listener, nChanges);
+        }
+
+
+        private Node mapDetached(DetachedNodeData detachedNode) {
+            Object linkedNode = detachedNode.getRefOrId();
+            if(linkedNode == null || linkedNode instanceof Long){
+                if(linkedNode == null) {
+                    if (keyPool == null) {
+                        linkedNode = graph.addNode(detachedNode.label());
+                    } else {
+                        linkedNode = graph.addNode(keyPool.next(), detachedNode.label());
+                    }
+                } else if (linkedNode instanceof Long) {
+                    linkedNode = graph.addNode(((Long) linkedNode).longValue(), detachedNode.label());
+                }
+                detachedNode.setRefOrId(linkedNode);
+                deferredInitializers.addLast(detachedNode);
+            }
+            return (Node) linkedNode;
+        }
```

Comment:
I find this rather hard to read. Not sure if I missed anything, but isn't this equivalent?
Alternatively we could break it up into two methods?

```suggestion
        private Node mapDetached(DetachedNodeData detachedNode) {
            Object linkedNode = detachedNode.getRefOrId();
            if (linkedNode == null) {
                if (keyPool == null) {
                    linkedNode = graph.addNode(detachedNode.label());
                } else {
                    linkedNode = graph.addNode(keyPool.next(), detachedNode.label());
                }
            } else if (linkedNode instanceof Long) {
                linkedNode = graph.addNode(((Long) linkedNode).longValue(), detachedNode.label());
            }

            detachedNode.setRefOrId(linkedNode);
            deferredInitializers.addLast(detachedNode);
            return (Node) linkedNode;
        }

```